src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVMKernels.cpp

```diff
@@ -663,6 +663,339 @@ INST_FluxKernel( 5, FaceElementStencil );
+           localIndex const stencilSize,
+           real64 const & dt,
+           arraySlice1d< localIndex const > const seri,
+           arraySlice1d< localIndex const > const sesri,
+           arraySlice1d< localIndex const > const sei,
+           arraySlice1d< real64 const > const stencilWeights,
+           ElementViewConst< arrayView1d< real64 const > > const & pres,
+           ElementViewConst< arrayView1d< real64 const > > const & gravCoef,
+           ElementViewConst< arrayView3d< real64 const > > const & phaseRelPerm,
+           ElementViewConst< arrayView3d< real64 const > > const & phaseVisc,
+           ElementViewConst< arrayView3d< real64 const > > const & phaseDens,
+           ElementViewConst< arrayView3d< real64 const > > const & phaseMassDens,
+           ElementViewConst< arrayView4d< real64 const > > const & phaseCompFrac,
+           ElementView< arrayView2d< real64 > > const & phaseOutflux,
+           ElementView< arrayView2d< real64 > > const & compOutflux )
+{
+  // loop over phases, compute and upwind phase flux and sum contributions to each component's flux
+  for( localIndex ip = 0; ip < numPhases; ++ip )
+  {
+    // clear working arrays
+    real64 densMean{};
+
+    // create local work arrays
+    real64 presGrad{};
+    real64 gravHead{};
+
+    // calculate quantities on primary connected cells
+    for( localIndex i = 0; i < NUM_ELEMS; ++i )
+    {
+      localIndex const er  = seri[i];
+      localIndex const esr = sesri[i];
+      localIndex const ei  = sei[i];
+
+      // average density across the face
+      densMean += 0.5 * phaseMassDens[er][esr][ei][0][ip];
+    }
+
+    //***** calculation of phase volumetric flux *****
+
+    // compute potential difference MPFA-style
+    for( localIndex i = 0; i < stencilSize; ++i )
+    {
+      localIndex const er  = seri[i];
+      localIndex const esr = sesri[i];
+      localIndex const ei  = sei[i];
+      real64 const weight  = stencilWeights[i];
+
+      presGrad += weight * pres[er][esr][ei];
+      gravHead += weight * densMean * gravCoef[er][esr][ei];
+    }
+
+    // *** upwinding ***
+
+    // compute phase potential gradient
+    real64 const potGrad = presGrad - gravHead;
+
+    // choose upstream cell
+    localIndex const k_up = (potGrad >= 0) ? 0 : 1;
+
+    localIndex const er_up  = seri[k_up];
+    localIndex const esr_up = sesri[k_up];
+    localIndex const ei_up  = sei[k_up];
+
+    real64 const mobility = phaseRelPerm[er_up][esr_up][ei_up][0][ip] / phaseVisc[er_up][esr_up][ei_up][0][ip];
+
+    // increment the phase (volumetric) outflux of the upstream cell
+    real64 const absPhaseFlux = fabs( dt * mobility * potGrad );
+    phaseOutflux[er_up][esr_up][ei_up][ip] += absPhaseFlux;
+
+    // increment the component (mass/molar) outflux of the upstream cell
+    for( localIndex ic = 0; ic < NC; ++ic )
+    {
+      compOutflux[er_up][esr_up][ei_up][ic] += phaseCompFrac[er_up][esr_up][ei_up][0][ip][ic]
```

Comment:
Do these sums need to be atomic?

Reply:
Yes! It is a mistake, I have not started testing this with threads yet. Thanks for the catch!